src/services/api/features/package-actions/services/seatool-write-service.ts

```diff
@@ -0,0 +1,101 @@
+import * as sql from "mssql";
+import * as query from "@/features/package-actions/queries";
+import { raiIssueSchema } from "shared-types";
+import { z } from "zod";
+import { APIError } from "./error-handle-service";
+
+type ISeatoolWriteService = {
+  withdrawRai: (queryParams: query.WithdrawRaiQueryParams) => Promise<void>;
+  withdrawPackage: (queryParams: query.WithdrawPackageParams) => Promise<void>;
+  issueRai: (raiData: z.infer<typeof raiIssueSchema>) => Promise<void>;
+  respondToRai: (queryParams: query.RespondToRaiQueryParams) => Promise<void>;
+  changePackageStatus: (
+    queryParams: query.ChangePackageStatusQueryParams
+  ) => Promise<void>;
+};
+
+export class SeatoolWriteService implements ISeatoolWriteService {
+  private readonly seatoolTransaction: sql.Transaction;
+
+  constructor(seatoolConnection: sql.ConnectionPool) {
+    this.seatoolTransaction = new sql.Transaction(seatoolConnection);
+  }
+
+  private handleTransaction = async (
+    transactionCallback: () => Promise<void>
+  ) => {
+    try {
+      await this.seatoolTransaction.begin();
+
+      await transactionCallback();
+
+      await this.seatoolTransaction.commit();
+    } catch (error: unknown) {
+      await this.seatoolTransaction.rollback();
+
+      console.error(error);
+
+      throw new APIError("Failed to write to Seatool.");
+    }
+  };
+
+  async withdrawRai({
+    id,
+    activeRaiDate,
+    withdrawnDate,
+  }: query.WithdrawRaiQueryParams) {
+    await this.handleTransaction(async () => {
+      await this.seatoolTransaction
+        .request()
+        .query(query.withdrawRaiQuery({ id, activeRaiDate, withdrawnDate }));
+    });
+  }
+
+  async withdrawPackage({
+    activeRaiDate,
+    withdrawnDate,
+    id,
+  }: query.WithdrawPackageParams) {
+    await this.handleTransaction(async () => {
+      await this.seatoolTransaction
+        .request()
+        .query(
+          query.withdrawPackageQuery({ activeRaiDate, withdrawnDate, id })
+        );
+    });
+  }
+
+  async issueRai(raiData: z.infer<typeof raiIssueSchema>) {
+    await this.handleTransaction(async () => {
+      await this.seatoolTransaction.request().query(
+        query.issueRaiQuery({
+          id: raiData.id,
+          requestedDate: raiData.requestedDate,
+        })
+      );
+    });
+  }
+
+  async respondToRai({
+    id,
+    latestRai,
+    responseDate,
+  }: query.RespondToRaiQueryParams) {
+    await this.handleTransaction(async () => {
+      await this.seatoolTransaction
+        .request()
+        .query(query.respondToRaiQuery({ id, latestRai, responseDate }));
+    });
+  }
+
+  async changePackageStatus({
+    id,
+    status,
+  }: query.ChangePackageStatusQueryParams) {
+    await this.handleTransaction(async () => {
+      await this.seatoolTransaction
+        .request()
+        .query(query.changePackageStatusQuery({ id, status }));
```

Comment:
What would be the selling points for having the `changePackageStatusQuery` abstraction?

Personally, when it comes to SQL queries, i try to avoid the any function abstractions:
```js
await this.handleTransaction(async () => {
    await this.seatoolTransaction
    .request()
    .query(`
    UPDATE SEA.dbo.State_Plan
    SET SPW_Status_ID = (Select SPW_Status_ID from SEA.dbo.SPW_Status where     SPW_Status_DESC = '${status}')
    WHERE ID_Number = '${id}'
    `
    );

```

Reply:
Ditto for the other methods above